Show desktop notifications for changes

alepop opened the pull request:
Resolve #253

alepop committed:
Add event handlers and some electron improvements

alepop committed:
Add Notify factory

alepop committed:
Add tests for Notify factory

alepop committed:
Merge branch 'development' into 253-add-desktop-notification

slaweet commented:
@alepop I tried to test this PR, but I couldn't get it to show any notification. I think it's related to the fact that we're using `window.requestAnimationFrame` (in sync service [0]) to get account updates, so it doesn't check server when the app window is not active.

[0] https://github.com/LiskHQ/lisk-nano/blob/development/src/services/sync.js

alepop commented:
@slaweet I fix this issue in commit 7677d26
```
+    webPreferences: {
 +      // Avoid app throttling when Electron is in background
 +      backgroundThrottling: false,
 +    },
```

slaweet commented:
@alepop thanks for the update. Now that I knew it should be working I found out that the notification was appearing on a different screen than the window was. But I suppose that's just my own settings.
So the feature works. 
Still, I'm not sure if this feature is worth all the requests that otherwise could have been spared while the app is in the background. @alihaghighatkhah , @karmacoma what do you think?
Maybe we can decrease the request interval when in background.

alihaghighatkhah commented:
@slaweet  Decreasing the sync broadcast rate could be a solution. I don't have anything against it.

alepop commented:
@slaweet, @alihaghighatkhah  what do you think the best interval in the inactive window `20000`, `30000` ?

slaweet commented:
I would go even with `60000`. 
In future, lisk core should provide some web socket API for this use case, so we don't have to choose arbitrary poll intervals.

alepop committed:
Change sync interval when app is not in focus

alihaghighatkhah commented on src/components/main/main.js:
```diff
@@ -17,7 +17,7 @@ app.component('main', {
   controller: class main {
     constructor($scope, $rootScope, $timeout, $q, $state, Peers,
-      dialog, SendModal, Account, AccountApi) {
+      dialog, SendModal, Account, AccountApi, Notify) {
```
Please change the name of the service to `Notification`.

alihaghighatkhah commented on src/components/main/main.js:
```diff
@@ -110,6 +111,10 @@ app.component('main', {
           }
 
+          this.notify.about(
+            'deposite',
```
Please correct the typo: deposite -> deposit. Please make sure it's fixed everywhere.

alihaghighatkhah commented on src/components/main/main.js:
```diff
@@ -110,6 +111,10 @@ app.component('main', {
             delete res.publicKey;
           }
 
+          this.notify.about(
```
Please check if `res.balance > this.account.get().balance` then call this method.

alepop replied:
I use subtraction because Notification display how many `LSK` received, while `res.balance` and `this.account.get().balance` display total balance amount.

alihaghighatkhah replied:
Yes, exactly. So you check the the balance is increased, then you pass the increased amount to `Notification.about`. My point is we should check and call if necessary, not always call and then perform if necessary.

alepop replied:
I understood you

alihaghighatkhah commented on src/services/notify.js:
```diff
@@ -0,0 +1,73 @@
+ * @module app
+ * @submodule Notify
+ */
+app.factory('Notify', ($window, lsk) => {
```
As mentioned before, Please change the name of the factory to `Notification`;

alihaghighatkhah commented on src/services/notify.js:
```diff
@@ -0,0 +1,73 @@
+     * @private
+     * @memberof Notify
+     */
+    __deposite(amount) { // eslint-disable-line
```
Please keep only one underscore sign and fix the typo. The method name must be `_deposit`.

alihaghighatkhah commented on src/services/notify.js:
```diff
@@ -0,0 +1,73 @@
+    __deposite(amount) { // eslint-disable-line
+      if (amount > 0) {
+        new $window.Notification( // eslint-disable-line
+          'LSK received',
```
Please remove the new line. simply use white space between the string and the brace.

alihaghighatkhah replied:
and since you're checking while calling this function in `main`, there's not need to check here any more.

alihaghighatkhah commented on src/index.pug:
```diff
@@ -7,6 +7,10 @@ html
       body {
         background-color: #eee !important;
       }
+    script.
```
Webpack has a built-in plugin named `ProvidePlugin`, you can use to create global variables. and since we're just using this for electron app, we can check the environment and add it only when building production package.

alepop replied:
Maybe are you talking about `DefinePlugin`? Because `ProvidePlugin` will load the module but not create a global variable.

alihaghighatkhah replied:
To the best of my knowledge `ProvidePlugin` can also do that. but sure, use which ever you find more fit. I just mean not to use script tags.

alepop replied:
I came to the conclusion that the use `window.require` in script tag is the simplest way to loading `electron`. Because of `DefinePlugin` use common `require` and it does not load electron to the browser.

alepop committed:
- Rename Notify -> Notification
- Fix typo
- Define `PRODUCTION` flag
- Load `ipcRenderer` to window object through the `preload` option

reyraa committed:
Merge branch 'development' into 253-add-desktop-notification

reyraa committed:
Merge branch 'development' into 253-add-desktop-notification

alepop committed:
Merge remote-tracking branch 'upstream/development' into 253-add-desktop-notification

reyraa committed:
Merge branch 'development' into 253-add-desktop-notification

reyraa commented:
Thank you @alepop
